crates/adapters/src/format/avro/deserializer.rs

```diff
@@ -41,211 +46,112 @@ use std::{
+            )));
+        };
+
+        Ok(SeqDeserializer {
             input: input.iter(),
-        }
+            item_schema: array_schema.items.as_ref(),
+        })
     }
 }
 
 impl<'de> MapDeserializer<'de> {
-    pub fn new(input: &'de HashMap<String, Value>) -> Self {
-        MapDeserializer {
+    pub fn new<E: de::Error>(
+        input: &'de HashMap<String, Value>,
+        schema: &'de Schema,
+    ) -> Result<Self, E> {
+        let Schema::Map(map_schema) = schema else {
+            return Err(de::Error::custom(format!(
+                "expected map schema, found {schema:?}"
+            )));
+        };
+
+        Ok(MapDeserializer {
+            value_schema: map_schema.types.as_ref(),
             input_keys: input.keys(),
             input_values: input.values(),
-        }
+        })
     }
 }
 
 impl<'de> RecordDeserializer<'de> {
-    pub fn new(input: &'de [(String, Value)]) -> Self {
-        RecordDeserializer {
+    pub fn new<E: de::Error>(
+        input: &'de [(String, Value)],
+        schema: &'de Schema,
+    ) -> Result<Self, E> {
+        let Schema::Record(record_schema) = schema else {
+            return Err(de::Error::custom(format!(
+                "expected record schema, found {schema:?}"
+            )));
+        };
+
+        Ok(RecordDeserializer {
+            record_schema,
             input: input.iter(),
             value: None,
-        }
-    }
-}
-
-impl<'a> EnumUnitDeserializer<'a> {
-    pub fn new(input: &'a str) -> Self {
-        EnumUnitDeserializer { input }
-    }
-}
-
-impl<'de> EnumDeserializer<'de> {
-    pub fn new(input: &'de [(String, Value)]) -> Self {
-        EnumDeserializer { input }
-    }
-}
-
-impl<'de> de::EnumAccess<'de> for EnumUnitDeserializer<'de> {
-    type Error = Error;
-    type Variant = Self;
-
-    fn variant_seed<V>(self, seed: V) -> Result<(V::Value, Self::Variant), Self::Error>
-    where
-        V: DeserializeSeed<'de>,
-    {
-        Ok((
-            seed.deserialize(StringDeserializer {
-                input: self.input.to_owned(),
-            })?,
-            self,
-        ))
-    }
-}
-
-impl<'de> de::VariantAccess<'de> for EnumUnitDeserializer<'de> {
-    type Error = Error;
-
-    fn unit_variant(self) -> Result<(), Error> {
-        Ok(())
-    }
-
-    fn newtype_variant_seed<T>(self, _seed: T) -> Result<T::Value, Error>
-    where
-        T: DeserializeSeed<'de>,
-    {
-        Err(de::Error::custom("unexpected Newtype variant"))
-    }
-
-    fn tuple_variant<V>(self, _len: usize, _visitor: V) -> Result<V::Value, Error>
-    where
-        V: Visitor<'de>,
-    {
-        Err(de::Error::custom("unexpected tuple variant"))
-    }
-
-    fn struct_variant<V>(
-        self,
-        _fields: &'static [&'static str],
-        _visitor: V,
-    ) -> Result<V::Value, Error>
-    where
-        V: Visitor<'de>,
-    {
-        Err(de::Error::custom("unexpected struct variant"))
+            value_index: 0,
```

Comment:
why zero?

Reply:
This is the index of the next field to deserialize, so it starts with 0.

Reply:
maybe it should be called next_field_index then

Reply:
done